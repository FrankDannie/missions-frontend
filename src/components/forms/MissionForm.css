@import './Form.css';

.mission-form {
  /* The .form-container class from Form.css will handle this */
}

.mission-form h2 {
  margin-bottom: 20px;
}

.mission-form label {
  display: block;
  margin-bottom: 10px;
}

.mission-form input,
.mission-form textarea {
  width: 100%;
  padding: 8px;
  margin-top: 5px;
  border: 1px solid #ccc;
  border-radius: 4px;
}

.mission-form button[type="submit"] {
  background-color: #007bff;
  color: #fff;
}

.mission-form button[type="button"] {
  background-color: #6c757d;
  color: #fff;
}
