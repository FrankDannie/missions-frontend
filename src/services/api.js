import axios from 'axios';

const api = axios.create({
  baseURL: process.env.REACT_APP_API_BASE_URL,
});

export const fetchRobots = () => api.get('/robots/');
export const fetchRobotById = (id) => api.get(`/robots/${id}/`); 
export const createRobot = (data) => api.post('/robots/', data);
export const updateRobot = (id, data) => api.put(`/robots/${id}/`, data);
