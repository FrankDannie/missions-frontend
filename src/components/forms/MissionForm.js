import React, { useState, useEffect } from 'react';
import {
  createMission,
  updateMission,
  fetchMissionById,
  fetchRobots,
} from '../../services/api';
import Labels from '../common/labels';
import './MissionForm.css';

const MissionForm = ({ missionId, onClose, onMissionUpdate }) => {
  const initialMissionState = { name: '', description: '', robot_id: '' };
  const [mission, setMission] = useState(initialMissionState);
  const [robots, setRobots] = useState([]);

  useEffect(() => {
    if (missionId) {
      fetchMissionById(missionId).then((response) => {
        setMission(response.data || initialMissionState);
      });
    } else {
      setMission(initialMissionState);
    }
    fetchRobots().then((response) => {
      setRobots(response.data || []);
    });
  }, [missionId]);

  const handleChange = (e) => {
    const { name, value } = e.target;
    setMission((prevMission) => ({ ...prevMission, [name]: value }));
  };

  const handleSubmit = (e) => {
    e.preventDefault();
    const payload = {
      name: mission.name,
      description: mission.description,
      robot_id: mission.robot_id,
    };
    if (missionId) {
      updateMission(missionId, payload).then(() => {
        onMissionUpdate();
        setMission(initialMissionState); // Reset the form fields
        onClose();
      });
    } else {
      createMission(payload).then(() => {
        onMissionUpdate();
        setMission(initialMissionState); // Reset the form fields
        onClose();
      });
    }
  };

  const handleCancel = () => {
    setMission(initialMissionState);
    onClose();
  };

  return (
    <div className="form-container mission-form">
      <h2>
        {missionId
          ? Labels.mission.UPDATE_MISSION
          : Labels.mission.CREATE_MISSION}
      </h2>
      <form onSubmit={handleSubmit}>
        <label>
          {Labels.mission.MISSION_NAME}
          <input
            type="text"
            name="name"
            value={mission.name}
            onChange={handleChange}
            required
          />
        </label>
        <label>
          {Labels.mission.MISSION_DESCRIPTION}
          <textarea
            name="description"
            value={mission.description}
            onChange={handleChange}
            required
          />
        </label>
        <label>
          {Labels.mission.ROBOT_SELECT}
          <select
            name="robot_id"
            value={mission.robot_id}
            onChange={handleChange}
            required
          >
            <option value="">{Labels.mission.SELECT_ROBOT}</option>
            {robots.map((robot) => (
              <option key={robot.id} value={robot.id}>
                {robot.name}
              </option>
            ))}
          </select>
        </label>
        <button type="submit">
          {missionId ? Labels.mission.UPDATE : Labels.mission.CREATE}
        </button>
        <button type="button" onClick={handleCancel}>
          {Labels.CANCEL}
        </button>
      </form>
    </div>
  );
};

export default MissionForm;
