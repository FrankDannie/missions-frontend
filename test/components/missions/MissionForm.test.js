import React from 'react';
import { render, screen, fireEvent, waitFor } from '@testing-library/react';
import '@testing-library/jest-dom/extend-expect'
import MissionForm from '../../../src/components/forms/MissionForm';
import { createMission, updateMission, fetchMissionById, fetchRobots } from '../../../src/services/api';
import Labels from '../../../src/components/common/labels';

jest.mock('../../../src/services/api');

describe('MissionForm', () => {
  const missionId = 1;
  const missionData = { name: 'Test Mission', description: 'Test Description', robot_id: '1' };
  const robotsData = [{ id: '1', name: 'Robot 1' }, { id: '2', name: 'Robot 2' }];

  beforeEach(() => {
    jest.clearAllMocks();
  });

  test('renders form for creating a new mission', async () => {
    fetchRobots.mockResolvedValue({ data: robotsData });

    const { asFragment } = render(<MissionForm onClose={() => {}} onMissionUpdate={() => {}} />);

    // Wait for robots to be fetched and check if they are in the dropdown
    await waitFor(() => expect(screen.getByText(/Robot 1/i)).toBeInTheDocument());

    // Verify the rendered output using snapshot
    expect(asFragment()).toMatchSnapshot();
  });

  test('renders form for updating an existing mission', async () => {
    fetchMissionById.mockResolvedValue({ data: missionData });
    fetchRobots.mockResolvedValue({ data: robotsData });

    const { asFragment } = render(<MissionForm missionId={missionId} onClose={() => {}} onMissionUpdate={() => {}} />);

    // Wait for mission data to be fetched and form fields to be populated
    await waitFor(() => {
      expect(screen.getByDisplayValue(/Test Mission/i)).toBeInTheDocument();
      expect(screen.getByDisplayValue(/Test Description/i)).toBeInTheDocument();
      expect(screen.getByDisplayValue(/1/i)).toBeInTheDocument();
    });

    // Verify the rendered output using snapshot
    expect(asFragment()).toMatchSnapshot();
  });

  test('submits the form to update an existing mission', async () => {
    fetchMissionById.mockResolvedValue({ data: missionData });
    fetchRobots.mockResolvedValue({ data: robotsData });
    updateMission.mockResolvedValue({});

    const onMissionUpdate = jest.fn();
    const onClose = jest.fn();

    render(<MissionForm missionId={missionId} onClose={onClose} onMissionUpdate={onMissionUpdate} />);

    // Wait for mission data to be fetched and form fields to be populated
    await waitFor(() => {
      expect(screen.getByDisplayValue(missionData.name)).toBeInTheDocument();
      expect(screen.getByDisplayValue(missionData.description)).toBeInTheDocument();
    });

    // Update the form fields
    fireEvent.change(screen.getByLabelText(Labels.mission.NAME), { target: { value: 'Updated Mission' } });
    fireEvent.change(screen.getByLabelText(Labels.mission.DESCRIPTION), { target: { value: 'Updated Description' } });

    // Submit the form
    fireEvent.click(screen.getByText(Labels.form.UPDATE));

    // Wait for the updateMission API call
    await waitFor(() => expect(updateMission).toHaveBeenCalledWith(missionId, {
      name: 'Updated Mission',
      description: 'Updated Description',
      robot_id: missionData.robot_id
    }));

    // Check if the callbacks were called
    expect(onMissionUpdate).toHaveBeenCalled();
    expect(onClose).toHaveBeenCalled();
  });


  
  test('handles cancel button click', () => {
    const onClose = jest.fn();

    render(<MissionForm onClose={onClose} onMissionUpdate={() => {}} />);

    // Click the cancel button
    fireEvent.click(screen.getByText(/Cancel/i));

    // Check if the onClose callback was called
    expect(onClose).toHaveBeenCalled();
  });
});
