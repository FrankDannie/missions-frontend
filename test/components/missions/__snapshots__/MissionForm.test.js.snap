// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MissionForm renders form for creating a new mission 1`] = `
<DocumentFragment>
  <div
    class="form-container mission-form"
  >
    <h2>
      Create Mission
    </h2>
    <form>
      <label>
        Mission Name:
        <input
          name="name"
          required=""
          type="text"
          value=""
        />
      </label>
      <label>
        Description:
        <textarea
          name="description"
          required=""
        />
      </label>
      <label>
        Robot:
        <select
          name="robot_id"
          required=""
        >
          <option
            value=""
          >
            Select a robot
          </option>
          <option
            value="1"
          >
            Robot 1
          </option>
          <option
            value="2"
          >
            Robot 2
          </option>
        </select>
      </label>
      <button
        type="submit"
      >
        Create
      </button>
      <button
        type="button"
      >
        Cancel
      </button>
    </form>
  </div>
</DocumentFragment>
`;

exports[`MissionForm renders form for updating an existing mission 1`] = `
<DocumentFragment>
  <div
    class="form-container mission-form"
  >
    <h2>
      Update Mission
    </h2>
    <form>
      <label>
        Mission Name:
        <input
          name="name"
          required=""
          type="text"
          value="Test Mission"
        />
      </label>
      <label>
        Description:
        <textarea
          name="description"
          required=""
        >
          Test Description
        </textarea>
      </label>
      <label>
        Robot:
        <select
          name="robot_id"
          required=""
        >
          <option
            value=""
          >
            Select a robot
          </option>
          <option
            value="1"
          >
            Robot 1
          </option>
          <option
            value="2"
          >
            Robot 2
          </option>
        </select>
      </label>
      <button
        type="submit"
      >
        Update
      </button>
      <button
        type="button"
      >
        Cancel
      </button>
    </form>
  </div>
</DocumentFragment>
`;
